src/lib/provable/crypto/foreign-twisted.ts

```diff
@@ -0,0 +1,309 @@
+import {
+  TwistedCurveParams,
+  createCurveTwisted,
+} from '../../../bindings/crypto/elliptic-curve.js';
+import type { Group } from '../group.js';
+import { ProvablePureExtended } from '../types/struct.js';
+import { AlmostForeignField, createForeignField } from '../foreign-field.js';
+import { CurveTwisted, Point } from '../gadgets/twisted-curve.js';
+import { Field3 } from '../gadgets/foreign-field.js';
+import { assert } from '../gadgets/common.js';
+import { Provable } from '../provable.js';
+import { provableFromClass } from '../types/provable-derivers.js';
+import { l2Mask, multiRangeCheck } from '../gadgets/range-check.js';
+import { Bytes } from '../bytes.js';
+
+// external API
+export { createForeignTwisted, ForeignTwisted };
+
+// internal API
+export { toPoint, FlexiblePoint };
+
+type FlexiblePoint = {
+  x: AlmostForeignField | Field3 | bigint | number;
+  y: AlmostForeignField | Field3 | bigint | number;
+};
+function toPoint({ x, y }: ForeignTwisted): Point {
+  return { x: x.value, y: y.value };
+}
+
+class ForeignTwisted {
+  x: AlmostForeignField;
+  y: AlmostForeignField;
+
+  /**
+   * Create a new {@link ForeignTwisted} from an object representing the (twisted) x and y coordinates.
+   *
+   * Note: Inputs must be range checked if they originate from a different field with a different modulus or if they are not constants. Please refer to the {@link ForeignField} constructor comments for more details.
+   *
+   * @example
+   * ```ts
+   * let x = new ForeignTwisted({ x: 1n, y: 1n });
+   * ```
+   *
+   * **Warning**: This fails for a constant input which does not represent an actual point on the curve.
```

Comment:
If the code aims to be generic, we must also ensure it is in the prime subgroup.
If the code is strictly for Curve25519, we should mention it.